### PR TITLE
MST Flow for asynchronous actions

### LINKED ISSUE #25: Improvement: Use MST `flow`
## Recommended Improvement

In our MobX State Tree Stores, when a method relies on an asynchronous action (e.g. "fetch Task"), we should define those methods using MST's `flow`

**The Problem**
- Our MST Stores like `MLAuthStore` have explicit setter functions, like `setStatus(val)`
- For preference, we should use `self.status = val` to reduce the number of unnecessary setter methods.
- Currently, `setStatus()` is _required_ because we cannot set the value of a Store's member, if we're changing said value from a nested function. This is pertinent in our asynchronous fetch methods, which define the 'on success' functions within the Promise chain.

e.g.
```
// Snapshot 1: Original
checkUser () {
  self.setStatus(ASYNC_STATES.FETCHING)
  oauth.checkCurrent()
  .then(user => {
    self.setStatus(ASYNC_STATES.SUCCESS)
    self.setUser(user)
  })
},
```

vs

```
// Snapshot 2: Doesn't work
checkUser () {
  self.status = ASYNC_STATES.FETCHING  // This is fine
  oauth.checkCurrent()
  .then(user => {
    self.status = ASYNC_STATES.SUCCESS  // This will break
    self.user = user  // This will break
  })
},
```

**Proposed Improvement**
- Use MST's `flow`, which uses a generator function specifically for handling asynchronous actions.
- More details: https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md#using-generators 
- Example with `checkCurrent()` in the monorepo: https://github.com/zooniverse/front-end-monorepo/blob/59805511c03c0b169219c3172bb6e02488bb5071/packages/app-project/stores/User.js#L28-L41

For the source, see PR #22 for the conversation with @eatyourgreens on this matter.

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview

Closes #25 
Requires #23 

This PR ATTEMPTS to use MobX State Tree's [flow](https://github.com/mobxjs/mobx-state-tree/blob/master/docs/async-actions.md#using-generators) mechanism for creating asynchronous actions.

Unfortunately, I've hit a stumbling block - any time I use `function* myFunc () {}`, the app will crash at runtime, with the error `Uncaught ReferenceError: regeneratorRuntime is not defined`

I suspect this is an issue with the way that the project is set up: either the Babel presets in `.babelrc` aren't working properly, or a Babel plugin isn't in place (e.g. `@babel/plugin-transform-regenerator`)

### Status

I'm putting this on hold for the moment, as I'm doing too much debugging to get an improvement working. I'd appreciate any advice anyone else has on this.

⚠️ WIP. Not yet functioning - Do not merge!

### CODE REVIEW

File: src/store/AuthStore.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      self.status = ASYNC_STATES.SUCCESS
+      self.user = user
+    })
+  }),
+
+  login () {
+    oauth.signIn(computeRedirectURL(window))
+  },
+
+  logout: flow(function* () {
```
Comment (eatyourgreens) [inline src/store/AuthStore.js]: I think you need to name the generator here.
```suggestion
  logout: flow(function * logout() {
```

Comment (eatyourgreens) [reply src/store/AuthStore.js]: At least, when I started working with generators, they wouldn't work with anonymous functions.

Comment (shaunanoordin) [reply src/store/AuthStore.js]: Thanks! I've just updated the code to add a name to the anon logout function, though I'm still encountering that darn `regenerator undefined` error.

Comment (shaunanoordin) [reply src/store/AuthStore.js]: Wait I didn't realise in GitHub you can make code changes directly in your comments, and that I can commit the suggestions from within the PR's comments. Sweet!

**Comment by shaunanoordin** [comment]
Edit: changed target base to #23 `status-errors` to make reviewing the MST-flow specific changes easier.

**Comment by eatyourgreens** [comment]
@shaunanoordin the latest commit should fix the generator error.

**Comment by eatyourgreens** [comment]
Working now for me:
Screenshot showing my Zooniverse account logged in to the prototype.

**Comment by shaunanoordin** [comment]
Confirmed! The `@babel/polyfill` solution works for me too! 🎉 

I'll proceed to update the other stores with `flow()` now that this is working.

**Comment by shaunanoordin** [comment]
## PR Update

This PR is now complete (all three stores with async functions use MST flows), thanks to the solution provided by @eatyourgreens.

To use MST's `flow`,
- `@babel/polyfill` needs to be set up for the project (this was the most WTF one)
- the generator `function *`s shouldn't be anonymous
- use `yield` instead of Promise chains.

⚠️ The [@babel/polyfill](https://babeljs.io/docs/en/babel-polyfill) solution is temporary, as it has been deprecated. A more permanent solution should be looked into later - possibly something to do with using something called `core-js/stable` and `regenerator-runtime/runtime ` ?

### Status

Ready for roll. 👍 

Mind you that this PR is merge-targeting #23 `status-errors`; that PR should be merged first, and then this PR be rebased and re-targeted to `master`

**Comment by shaunanoordin** [comment]
#23 has been merged, and this has been retargeted. 👍

Let me fix those conflicts.

**Comment by shaunanoordin** [comment]
👌 Awesome, all conflicts resolved. I'm going to go ahead and merge this - thanks for the review!

**Pull request merged**